### DISCUSSION TITLE
process.cwd() Different paths in development and production environment

### DISCUSSION THREAD

**Discussion by SabrioBiblio** [vercel/vercel#11197] [Help] [answered]
### Question

I'm using next js 14.

In each page folder I have a file (**meta.md**) that I read using fs. Let it be written accordingly to these files
`const postsDirectoryPath = process.cwd();`

In the development environment, everything works and the files are read, this is understandable, since **process.cwd** refers to the root directory of the project. But on the vercel, **process.cwd** refers to .next, where naturally the **meta.md** file no longer exists. I have some problems with debugging on Vercel and I can't keep track of everything.

But there is some confusion about how this works in production, since I call my function that reads the files in the top **page.tsx** (homepage) file and it surprisingly works. Doesn't work if the md file reading function is called at levels below (another pages in another folders)

**Comment by SabrioBiblio** [comment]
I see that this is a common problem related to verceland the file system in production. I tried to implement this in different ways, but it doesn't seem to work

**Comment by GrishonNganga** [comment]
Don't know if any of you was able to fix the issue. I managed to go around the issue by putting the .md files  I needed to open in the lib folder. Turns out vercel always includes the lib folder as part of the assets under src/lib. That way I was able to open the files with process.cwd() both locally and on vercel.

**Comment by lob0codes** [comment]
I was struggling with a similar issue. Not sure if the solution found will work for you.

If you render something like "<p>{process.cwd()}</p>" in prod (app deployed in Vercel), it renders the directory path "var/task".
Which as expected, is different from what one will get in a local environment (npm run dev).

So, I tried to render something like this:

const filesDirectoryList = fs.readdir(process.cwd())

{filesDirectoryList.map(file => (<p>{file}</p>)}

And got a list of files where "public" directory was not visible. 
So, after looking for a solution, I saw a proposal of adding "public" to the desired directory path: 

import fs from "fs";
import path from "path";

const filePath = path.join(process.cwd(), "public", "example.md");
const content = fs.readFileSync(filePath);

You may add a more complex hierarchy of folders like "posts/example.md" and it also works.

It did the job, I was able target the desired files.

**Comment by leerob** [comment]
There's some more details about this files get included here: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

Do do you have a small example of the issue you're seeing?

**Comment by SabrioBiblio** [comment, marked as answer]
I solved this issue.
Everything is quite logical.
However, I don’t quite remember the decision, since I haven’t touched the project for more than two months.
But here is an example with posts that I implemented

```javascript
/*The posts folder is in the root of the project*/
const pathToPosts = process.env.NODE_ENV === "development" ? "/posts" : "posts";
const postsDirectoryPath = process.cwd();
    const files = await fs.readdir(path.join(postsDirectoryPath, pathToPosts), {
      withFileTypes: true,
    });
```

**Comment by thebiltheory** [comment]
Might also be [related to this](https://github.com/vercel/next.js/issues/8251). 

If you're using `next`, using `turbotrace` did the trick for me. 

See [the official documentation](https://nextjs.org/docs/pages/api-reference/next-config-js/output#experimental-turbotrace) 

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: "standalone",
  experimental: {
    turbotrace: {},
  },
};
```

EDIT: `processCwd: process.cwd()`  wasn't needed.

**Comment by rccyx** [comment]
Running Next 14. 
I get the same problem with `process.cwd()`, where I have this setup to retrieve files from the blog directory @ `/public/blog`
````ts
export async function getBlogPosts(
  blogDirectory: string = BLOG_CONTENT_PATH
): Promise<Optional<PostData[]>> {
  console.log(
    'Invoked all blog posts from: ' + path.join(process.cwd(), blogDirectory)
  );
  return getMDXData(path.join(process.cwd(), blogDirectory));
}
````

I get this on vercel 
````
Invoked all blog posts from: /var/task/public/blogs
Invoked all mdx data from: /var/task/public/blogs
Invoked all blog posts from: /var/task/public/blogs
Invoked all mdx data from: /var/task/public/blogs
Error reading MDX files: [Error: ENOENT: no such file or directory, scandir '/var/task/public/blogs'] {
errno: -2,
code: 'ENOENT',
syscall: 'scandir',
path: '/var/task/public/blogs'
}
Error reading MDX files: [Error: ENOENT: no such file or directory, scandir '/var/task/public/blogs'] {
errno: -2,
code: 'ENOENT',
syscall: 'scandir',
path: '/var/task/public/blogs'
}
````

Local production build
````
Invoked all blog posts from: /home/ashgw/work/ashgw.me/public/blogs
Invoked all mdx data from: /home/ashgw/work/ashgw.me/public/blogs
Invoked all blog posts from: /home/ashgw/work/ashgw.me/public/blogs
Invoked all mdx data from: /home/ashgw/work/ashgw.me/public/blogs
````
This solved it
````ts
/** @type {import('next').NextConfig} */

const nextConfig = {
  experimental: {
    outputFileTracingIncludes: {
      '/blog': ['./public/**/*'],
    },
  },
}
````

**Comment by Demi1024** [comment]
> Running Next 14. I get the same problem with `process.cwd()`, where I have this setup to retrieve files from the blog directory @ `/public/blog`
> 
> ```ts
> export async function getBlogPosts(
>   blogDirectory: string = BLOG_CONTENT_PATH
> ): Promise<Optional<PostData[]>> {
>   console.log(
>     'Invoked all blog posts from: ' + path.join(process.cwd(), blogDirectory)
>   );
>   return getMDXData(path.join(process.cwd(), blogDirectory));
> }
> ```
> 
> I get this on vercel
> 
> ```
> Invoked all blog posts from: /var/task/public/blogs
> Invoked all mdx data from: /var/task/public/blogs
> Invoked all blog posts from: /var/task/public/blogs
> Invoked all mdx data from: /var/task/public/blogs
> Error reading MDX files: [Error: ENOENT: no such file or directory, scandir '/var/task/public/blogs'] {
> errno: -2,
> code: 'ENOENT',
> syscall: 'scandir',
> path: '/var/task/public/blogs'
> }
> Error reading MDX files: [Error: ENOENT: no such file or directory, scandir '/var/task/public/blogs'] {
> errno: -2,
> code: 'ENOENT',
> syscall: 'scandir',
> path: '/var/task/public/blogs'
> }
> ```
> 
> Local production build
> 
> ```
> Invoked all blog posts from: /home/ashgw/work/ashgw.me/public/blogs
> Invoked all mdx data from: /home/ashgw/work/ashgw.me/public/blogs
> Invoked all blog posts from: /home/ashgw/work/ashgw.me/public/blogs
> Invoked all mdx data from: /home/ashgw/work/ashgw.me/public/blogs
> ```
> 
> This solved it
> 
> ```ts
> /** @type {import('next').NextConfig} */
> 
> const nextConfig = {
>   experimental: {
>     outputFileTracingIncludes: {
>       '/blog': ['./public/**/*'],
>     },
>   },
> }
> ```
Works for me, thanks 😊
